### PR TITLE
Migrate builds tests away from using sample imagestreams since samples will be optional

### PR DESCRIPTION
**Pull request by bparees**
Resolving some of the failures seen in:
https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-ci-4.11-e2e-aws-no-capabilities/1503711956505202688

which runs our e2e but with optional capabilities (in this case, samples operator)
disabled.

Note, in a few cases we used "foo:latest" as the imagestream reference which had the nice property that it protected us as the imagestream version (e.g. nodejs:16) moved on to, say, nodejs:17.  None of these tests really care about the underlying image they are using, but it does mean the tests will no longer be automatically picking up newer versions of images as they become available, for better or worse.

### CODE REVIEW

**Comment by bparees** [comment]
@adambkaplan @gabemontero @coreydaley ptal, we need to move these tests away from relying on the samples operator asap so we can bring online our new e2e job that tests the cluster when optional capabilities are disabled

there will be a few more of these, but these were the easiest ones.

**Comment by wking** [comment]
For context on samples becoming optional, see openshift/cluster-samples-operator#414 and openshift/enhancements#922

**Review by gabemontero — CHANGES_REQUESTED** [review]
Don't you need to update https://github.com/openshift/origin/blob/master/test/extended/util/image/image.go#L15-L71 per the process defined at https://github.com/openshift/origin/tree/master/test/extended/util/image @bparees ?

Otherwise the changes here are OK.  I'm assuming you have already scanned for `DockerImage` from's with `image-registry.openshift-image-registry.svc:5000` or from's of type `ImageStream`/`ImageStreamTag` and these are the hits you got.

**Comment by bparees** [comment]
> Don't you need to update https://github.com/openshift/origin/blob/master/test/extended/util/image/image.go#L15-L71 per the process defined at https://github.com/openshift/origin/tree/master/test/extended/util/image @bparees ?

hrrrmm..  i'd forgotten about that.  That's going to be a problem since we're not going to mirror redhat registry content to that quay repo.  Though i'll also note that these tests would never have worked in a disconnected environment anyway, unless someone mirrored the content so the imagestreams could import it.  i'll need to think about what we need to do here.  It's also slightly complicated by the cases where the image pullspec is embedded in a yaml file, not in code where we can easily invoke LocationFor.

I'll have to see if i can dig up centos versions of these images that we can mirror, and then sort out what to do about the dynamic LocationFor substituting in the yaml files.

/hold

> Otherwise the changes here are OK. I'm assuming you have already scanned for DockerImage from's with image-registry.openshift-image-registry.svc:5000 or from's of type ImageStream/ImageStreamTag and these are the hits you got.

No, we actually have an e2e job running already w/ the samples operator disabled, so this was based on which tests failed in that job, linked in the PR description (followed by me digging into the failures to confirm they were related to not having the imagestreams/samples-operator present)

**Comment by gabemontero** [comment]
> > Don't you need to update https://github.com/openshift/origin/blob/master/test/extended/util/image/image.go#L15-L71 per the process defined at https://github.com/openshift/origin/tree/master/test/extended/util/image @bparees ?
> 
> hrrrmm.. i'd forgotten about that. That's going to be a problem since we're not going to mirror redhat registry content to that quay repo. Though i'll also note that these tests would never have worked in a disconnected environment anyway, unless someone mirrored the content so the imagestreams could import it. i'll need to think about what we need to do here. It's also slightly complicated by the cases where the image pullspec is embedded in a yaml file, not in code where we can easily invoke LocationFor.
> 
pretty sure that we discovered during https://github.com/openshift/origin/pull/26149 that Clayton's "bindata related libraries" handle the embedded yaml stuff.  Somewhere in that bindata loading code he has a LocationFor.

and I thought there was some mirroring of sample content going on, but I'm less sure about that one .... pretty sure about the bindata / LocationFor stuff.

> I'll have to see if i can dig up centos versions of these images that we can mirror, and then sort out what to do about the dynamic LocationFor substituting in the yaml files.
> 
> /hold
> 
> > Otherwise the changes here are OK. I'm assuming you have already scanned for DockerImage from's with image-registry.openshift-image-registry.svc:5000 or from's of type ImageStream/ImageStreamTag and these are the hits you got.
> 
> No, we actually have an e2e job running already w/ the samples operator disabled, so this was based on which tests failed in that job, linked in the PR description (followed by me digging into the failures to confirm they were related to not having the imagestreams/samples-operator present)

That's probably OK at this point.  I vaguely recall the switch from docker.io being more complicated  / iterative, but its possible the groundwork from that and Clayton's prior overhaul toward achieving disconnected "made things simpler".

Also, I'm assuming the image eco suite is off on purpose since samples are disabled.  That simplifies things too.

I'm sure you all are good to go wrt this.

**Comment by bparees** [comment]
hmmm, i'm not seeing anything that would be that aware.  It would have to understanding what part of the yaml contained an image reference that needed to be LocationFor'd.

I think things like this are simply bypassing the LocationFor logic and pulling directly from the location indicated (which will work, as long as the cluster under test is not disconnected):
https://github.com/openshift/origin/blob/1c7bf01e164c265713888e73589dd437a61eb000/test/extended/testdata/builds/build-postcommit/sti.yaml#L23

The fact that this PR mostly passed testing:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/26909/pull-ci-openshift-origin-master-e2e-gcp-builds/1503875911261360128

despite the fact that images like `registry.redhat.io/ubi8/php-74:latest` are not currently mirrored into `quay.io/openshift/community-e2e-images` proves the point, i think.

If anything was doing automated LocationFor substitution of those yaml values, the test would have ended up referencing the image on `quay.io/openshift/community-e2e-images` and then failing to pull the image, but the test passed.

Using imagestream values, or pullspecs that effectively point to imagestreams in the cluster, somewhat dodges the problem by giving the cluster an opportunity to redirect where the image is actually coming from.  So that is being lost by my changes here.  But i think there are deeper issues w/ the fact that fundamentally these tests(before or after my changes) do not align w/ the claimed rule that all images e2e's depend on, live in that community repository.

i'll keep digging.

**Comment by bparees** [comment]
there also seems to be at least some tests that are not actually conformant to this rule today:
https://github.com/openshift/origin/blob/1c7bf01e164c265713888e73589dd437a61eb000/test/extended/cli/builds.go#L87

**Comment by bparees** [comment]
I spoke with Derek about this and the "e2es must consume mirrored images from community-e2e" policy is specific to `conformance` e2e tests.  I'll double check that none of these are marked conformance and if that's true, i think we're ok to proceed

**Comment by bparees** [comment]
yeah none of these are `[Conformance]` tests so i think we're ok.  @gabemontero you want to give it the lgtm?

/hold cancel

**Comment by gabemontero** [comment]
> hmmm, i'm not seeing anything that would be that aware. It would have to understanding what part of the yaml contained an image reference that needed to be LocationFor'd.
> 
> I think things like this are simply bypassing the LocationFor logic and pulling directly from the location indicated (which will work, as long as the cluster under test is not disconnected):
> 
> https://github.com/openshift/origin/blob/1c7bf01e164c265713888e73589dd437a61eb000/test/extended/testdata/builds/build-postcommit/sti.yaml#L23
> 
> The fact that this PR mostly passed testing: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/26909/pull-ci-openshift-origin-master-e2e-gcp-builds/1503875911261360128
> 
> despite the fact that images like `registry.redhat.io/ubi8/php-74:latest` are not currently mirrored into `quay.io/openshift/community-e2e-images` proves the point, i think.
> 
> If anything was doing automated LocationFor substitution of those yaml values, the test would have ended up referencing the image on `quay.io/openshift/community-e2e-images` and then failing to pull the image, but the test passed.
> 
> Using imagestream values, or pullspecs that effectively point to imagestreams in the cluster, somewhat dodges the problem by giving the cluster an opportunity to redirect where the image is actually coming from. So that is being lost by my changes here. But i think there are deeper issues w/ the fact that fundamentally these tests(before or after my changes) do not align w/ the claimed rule that all images e2e's depend on, live in that community repository.
> 
> i'll keep digging.

Yeah I could be misrembering, and trying to retrace the history in that PR, and looking at the code again, I'm not finding anything definitive.  Can't shake the recollection though ...

... that said, given your other comments, about the other "danglers" and the multitude of ways images can be references with build, I'm fine with bypassing additions to https://github.com/openshift/origin/blob/master/test/extended/util/image/image.go#L15-L71 in this PR if that is your preference.

/lgtm

**Comment by bparees** [comment]
attempting to clarify the policy here:
https://github.com/openshift/origin/pull/26912

**Comment by gabemontero** [comment]
/lgtm

even  though @bparees added it as well

**Pull request merged**
